identity-apps-core/apps/recovery-portal/src/main/webapp/password-recovery-otp.jsp

```diff
@@ -0,0 +1,315 @@
+            return;
+        }
+        List<UserClaim> userClaims = new ArrayList<UserClaim>();
+
+        // Get the username claim string for the tenant
+        UserClaim userNameClaim = new UserClaim();
+        userNameClaim.setUri(USERNAME_CLAIM_URI_FOR_TENANT);
+        userNameClaim.setValue(MultitenantUtils.getTenantAwareUsername(username));
+        userClaims.add(userNameClaim);
+
+        String flawConfirmationCode = "";
+        String recoveryCode = "";
+        String channelId = "";
+        String screenValue = "";
+
+        try {
+            // Initiate password recovery.
+            RecoveryInitRequest recoveryInitRequest = new RecoveryInitRequest();
+            recoveryInitRequest.setClaims(userClaims);
+            Map<String, String> requestHeaders = new HashedMap();
+            if (request.getParameter("g-recaptcha-response") != null) {
+                requestHeaders.put("g-recaptcha-response", request.getParameter("g-recaptcha-response"));
+            }
+            List<AccountRecoveryType> resp = 
+                recoveryApiV2.initiatePasswordRecovery(recoveryInitRequest, tenantDomain, requestHeaders);
+            if (resp == null) {
+                /** Handle invalid username scenario. proceeds to next level without warning to 
+                avoid an attacker bruteforcing to learn the usernames. */
+                request.setAttribute("screenValue", "******" + getRandomNumberString(4, username));
+                request.setAttribute("resendCode", UUID.randomUUID().toString());
+                request.setAttribute("flowConfirmationCode", UUID.randomUUID().toString());
+                request.getRequestDispatcher("sms-otp.jsp").forward(request, response);
+                return;
+            }
+            for(AccountRecoveryType recoveryType: resp) {
+                if (recoveryType.getMode().equals("recoverWithNotifications")) {
+                    RecoveryChannelInformation channelInfo = recoveryType.getChannelInfo();
+                    recoveryCode = channelInfo.getRecoveryCode();
+                    List<RecoveryChannel> channels = channelInfo.getChannels();
+                    for(RecoveryChannel ch: channels) {
+                        flawConfirmationCode = recoveryType.getFlowConfirmationCode();
+                        if (ch.getType().equals(targetChannel)) {
+                            channelId = ch.getId();
+                            screenValue = ch.getValue();
+                            break;
+                        }
+                    }
+                }
+                if (StringUtils.isNotBlank(channelId)) {
+                    break;
+                }
+            }
+
+            /**
+             * Manage user don't have phone number set up in the account.
+             */
+            if (StringUtils.isBlank(channelId)) {
+                String recoveryPageQueryString = request.getParameter("urlQuery");
```

Comment:
Let's url encode the getParam calls to avoid xss exploitations. Check other places as well.

Reply:
fixed with https://github.com/wso2/identity-apps/pull/5817/commits/c4711e788c5f3c08e1e2906f20e2217b381f34e9